### PR TITLE
WIP add get_components method

### LINKED ISSUE #3536: ENH: make _get_ica_map a public method of ICA
We ever so often get people wondering how to retrieve ICA maps as numpy arrays, and I keep referring them to mne.preprocessing.ica._get_ica_map. It would probably be easier to make it a public method of ICA. Opinions?

### PR DESCRIPTION
**Pull request by jona-sassenhagen**
Remake of #3573 cause we had a lot of discussion.

closes #3536

@agramfort @mmagnuski @agostontorok

### CODE REVIEW

File: mne/preprocessing/ica.py [hunk 1]
```diff
@@ -662,6 +662,18 @@ def _transform_evoked(self, evoked):
...
+        """Get ICA topomap for components as numpy arrays
+
+        Returns
+        -------
+        components : array (n_channels, n_components)
+            The ICA components (maps).
+        """
+        fast_dot = _get_fast_dot()
+        return fast_dot(self.mixing_matrix_[:, :self.n_components_].T,
+                        self.pca_components_[:self.n_components_]).T
```
Comment (dengemann) [inline mne/preprocessing/ica.py]: what's the reason you transpose here?

Comment (jona-sassenhagen) [inline mne/preprocessing/ica.py]: @agramfort suggested it in #3573 ...

Comment (dengemann) [inline mne/preprocessing/ica.py]: ok.

**Comment by jona-sassenhagen** [comment]
Tbh I don't actually believe it's *that* important ...

**Comment by jona-sassenhagen** [comment]
This isn't important right?

```
======================================================================
ERROR: Test mne make_scalp_surfaces.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/miniconda/envs/testenv/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/travis/build/mne-tools/mne-python/mne/utils.py", line 819, in dec
    return function(*args, **kwargs)
  File "/home/travis/miniconda/envs/testenv/lib/python2.7/site-packages/numpy/testing/decorators.py", line 146, in skipper_func
    return f(*args, **kwargs)
  File "/home/travis/build/mne-tools/mne-python/mne/commands/tests/test_commands.py", line 142, in test_make_scalp_surfaces
    mne_make_scalp_surfaces.run()
  File "/home/travis/build/mne-tools/mne-python/mne/commands/mne_make_scalp_surfaces.py", line 61, in run
    options.no_decimate, options.verbose)
  File "<string>", line 2, in _run
  File "/home/travis/build/mne-tools/mne-python/mne/utils.py", line 702, in verbose
    return function(*args, **kwargs)
  File "/home/travis/build/mne-tools/mne-python/mne/commands/mne_make_scalp_surfaces.py", line 118, in _run
    n_triangles=n_tri)
  File "/home/travis/build/mne-tools/mne-python/mne/surface.py", line 817, in decimate_surface
    return _decimate_surface(points, triangles, reduction)
  File "/home/travis/build/mne-tools/mne-python/mne/surface.py", line 780, in _decimate_surface
    src = tvtk.PolyData(points=points, polys=triangles)
  File "/home/travis/miniconda/envs/testenv/lib/python2.7/site-packages/tvtk/tvtk_classes.zip/tvtk_classes/tvtk_helper.py", line 1745, in <lambda>
    PolyData = property(lambda self: get_class('PolyData'))
  File "/home/travis/miniconda/envs/testenv/lib/python2.7/site-packages/tvtk/tvtk_classes.zip/tvtk_classes/tvtk_helper.py", line 36, in get_class
    mod = get_module(fname)
  File "/home/travis/miniconda/envs/testenv/lib/python2.7/site-packages/tvtk/tvtk_classes.zip/tvtk_classes/tvtk_helper.py", line 28, in get_module
    mod = __import__('tvtk.tvtk_classes.%s'%fname, globals(), locals(), [fname])
  File "/home/travis/miniconda/envs/testenv/lib/python2.7/site-packages/tvtk/tvtk_classes.zip/tvtk_classes/poly_data.py", line 12, in <module>
  File "/home/travis/miniconda/envs/testenv/lib/python2.7/site-packages/tvtk/array_handler.py", line 28, in <module>
    from tvtk.array_ext import set_id_type_array
ImportError: numpy.core.multiarray failed to import
-------------------- >> begin captured stdout << ---------------------
```

**Comment by dengemann** [comment]
> This isn't important right?

seems unrelated, yes.

**Comment by jona-sassenhagen** [comment]
Travis error is the same as above.

**Comment by jona-sassenhagen** [comment]
Examples look fine here, good to go from my end.

**Comment by dengemann** [comment]
@jona-sassenhagen wanna keep the doc part for late? What about API reference/what's new?

**Comment by agramfort** [comment]
code LGTM but it needs a tutorial update + what's new

**Comment by jona-sassenhagen** [comment]
How about this?

File: tutorials/plot_artifacts_correction_ica.py [hunk 2]
```diff
@@ -236,20 +236,30 @@
...
 # Nice, we have found similar ICs from the other (simulated) runs!
-# This is even nicer if we have 20 or 100 ICA solutions in a list.
+# In this way, you can detect a type of artifact semi-automatically for example
+# for all subjects in a study.
+# The detected template can also be retrieved as an array and stored; this
+# array can be used as an alternative template to
+# :func:`mne.preprocessing.corrmap`.
+eog_component = reference_ica.get_components()[:, eog_inds[0]]
+
+# If you calculate a new ICA solution, you can provide this array instead of
```
Comment (mmagnuski) [inline tutorials/plot_artifacts_correction_ica.py]: I find this not very clear - why "calculate a new ICA solution"? You pass the template to corrmap, not ica, right?

File: tutorials/plot_artifacts_correction_ica.py [hunk 3]
```diff
@@ -236,20 +236,30 @@
...
+# specifying the template in reference to the list of ICA objects you want
```
Comment (mmagnuski) [inline tutorials/plot_artifacts_correction_ica.py]: maybe "... template as a reference to ..."?

File: tutorials/plot_artifacts_correction_ica.py (hunk 2 above)
Comment (jona-sassenhagen) [inline tutorials/plot_artifacts_correction_ica.py]: If you calculate a new ICA solution, the index possibly won't point to the correct component anymore. Storing the component allows you to bypass identifying the template component.

File: tutorials/plot_artifacts_correction_ica.py (hunk 3 above)
Comment (jona-sassenhagen) [inline tutorials/plot_artifacts_correction_ica.py]: No, that's something different? The template as an index is a reference to one of the components in one of the ICA objects.

File: tutorials/plot_artifacts_correction_ica.py [hunk 4]
```diff
@@ -236,20 +236,30 @@
...
+# Now we can run the CORRMAP algorithm.
 fig_template, fig_detected = corrmap(icas, template=template, label="blinks",
                                      show=True, threshold=.8, ch_type='mag')
 
 ###############################################################################
 # Nice, we have found similar ICs from the other (simulated) runs!
-# This is even nicer if we have 20 or 100 ICA solutions in a list.
+# In this way, you can detect a type of artifact semi-automatically for example
+# for all subjects in a study.
+# The detected template can also be retrieved as an array and stored; this
```
Comment (mmagnuski) [inline tutorials/plot_artifacts_correction_ica.py]: another nitpick: I wouldn't call it "detected template" (cause it was used in the detection but not detected itself) - maybe "chosen template" or simply just "template"

**Comment by mmagnuski** [comment]
Just a few nitpicks from my side.

File: tutorials/plot_artifacts_correction_ica.py (hunk 2 above)
Comment (mmagnuski) [inline tutorials/plot_artifacts_correction_ica.py]: Ok, I get it now. What about change it a little then: "Later, when you calculate another ICA solution for similar data ..."?

File: tutorials/plot_artifacts_correction_ica.py (hunk 3 above)
Comment (mmagnuski) [inline tutorials/plot_artifacts_correction_ica.py]: Yes, I get it, I just had a bad english intuition :)

**Comment by agramfort** [comment]
you need to rebase

**Comment by jona-sassenhagen** [comment]
Otherwise ok @agramfort ?

File: mne/preprocessing/ica.py [hunk 5]
```diff
@@ -662,6 +662,18 @@ def _transform_evoked(self, evoked):
 
         return sources
 
+    def get_components(self):
+        """Get ICA topomap for components as numpy arrays.
+
+        Returns
+        -------
+        components : array (n_channels, n_components)
```
Comment (agramfort) [inline mne/preprocessing/ica.py]: components : array, shape (n_channels, n_components)

**Comment by agramfort** [comment]
this needs rebase

**Comment by agramfort** [comment]
@jona-sassenhagen you need to rebase

**Comment by agramfort** [comment]
also set the PR to MRG if you're done

**Comment by agramfort** [comment]
thx @jona-sassenhagen 

i'll fix directly in master my nitpick

**Comment by jona-sassenhagen** [comment]
Thanks ... sorry, I was too distracted to finish this in time :(

**Pull request merged**